src/PowerShellEditorServices/Session/PromptNest.cs

```diff
@@ -0,0 +1,564 @@
+            {
+                return null;
+            }
+
+            // Typically we want to run PSReadLine on the current nest frame.
+            // The exception is when the current frame is remote, in which
+            // case we need to run it in it's own frame because we can't take
+            // over a remote pipeline through event invocation.
+            if (NestedPromptLevel > 1 && !IsRemote)
+            {
+                return CurrentFrame.PowerShell;
+            }
+
+            return isReadLine ? _readLineFrame.PowerShell : CurrentFrame.PowerShell;
+        }
+
+        /// <summary>
+        /// Get the <see cref="RunspaceHandle" /> for the current <see cref="PromptNestFrame" />.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken" /> that can be used to cancel the request.
+        /// </param>
+        /// <param name="isReadLine">Indicates whether this is for a PSReadLine command.</param>
+        /// <returns>The <see cref="RunspaceHandle" /> for the current frame.</returns>
+        internal RunspaceHandle GetRunspaceHandle(CancellationToken cancellationToken, bool isReadLine)
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            // Also grab the main runspace handle if this is for a ReadLine pipeline and the runspace
+            // is in process.
+            if (isReadLine && !_powerShellContext.IsCurrentRunspaceOutOfProcess())
+            {
+                GetRunspaceHandleImpl(cancellationToken, isReadLine: false);
+            }
+
+            return GetRunspaceHandleImpl(cancellationToken, isReadLine);
+        }
+
+
+        /// <summary>
+        /// Get the <see cref="RunspaceHandle" /> for the current <see cref="PromptNestFrame" />.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken" /> that will be checked prior to
+        /// completing the returned task.
+        /// </param>
+        /// <param name="isReadLine">Indicates whether this is for a PSReadLine command.</param>
+        /// <returns>
+        /// A <see cref="Task{RunspaceHandle}" /> object representing the asynchronous operation.
+        /// The <see cref="Task{RunspaceHandle}.Result" /> property will return the
+        /// <see cref="RunspaceHandle" /> for the current frame.
+        /// </returns>
+        internal async Task<RunspaceHandle> GetRunspaceHandleAsync(CancellationToken cancellationToken, bool isReadLine)
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            // Also grab the main runspace handle if this is for a ReadLine pipeline and the runspace
+            // is in process.
+            if (isReadLine && !_powerShellContext.IsCurrentRunspaceOutOfProcess())
+            {
+                await GetRunspaceHandleImplAsync(cancellationToken, isReadLine: false);
```

Comment:
do you need to return here?

Reply:
No, when PSRL is running in the main runspace it controls the pipeline thread. So the ReadLine handle and the main thread handle need to be obtained.